VoodooInput/VoodooInputSimulator/VoodooInputSimulatorDevice.cpp

```diff
@@ -214,53 +213,54 @@ void VoodooInputSimulatorDevice::constructReportGated(const VoodooInputEvent& mu
     }
     
     if (!input_active) {
         for (int i = 0; i < 15; i++) {
             touch_state[i] = 2;
         }
-
-        input_report.FINGERS[0].Size = 0x0;
-        input_report.FINGERS[0].Pressure = 0x0;
-        input_report.FINGERS[0].Touch_Major = 0x0;
-        input_report.FINGERS[0].Touch_Minor = 0x0;
+    
+        input_report->FINGERS[0].Size = 0x0;
+        input_report->FINGERS[0].Pressure = 0x0;
+        input_report->FINGERS[0].Touch_Major = 0x0;
+        input_report->FINGERS[0].Touch_Minor = 0x0;
         milli_timestamp += 10;
-        input_report.timestamp_buffer[0] = (milli_timestamp << 0x3) | 0x4;
-        input_report.timestamp_buffer[1] = (milli_timestamp >> 0x5) & 0xFF;
-        input_report.timestamp_buffer[2] = (milli_timestamp >> 0xd) & 0xFF;
-        handleReport(kernel_buffer, kIOHIDReportTypeInput);
+        input_report->timestamp_buffer[0] = (milli_timestamp << 0x3) | 0x4;
+        input_report->timestamp_buffer[1] = (milli_timestamp >> 0x5) & 0xFF;
+        input_report->timestamp_buffer[2] = (milli_timestamp >> 0xd) & 0xFF;
+        handleReport(input_report_buffer,kIOHIDReportTypeInput);
 
-        input_report.FINGERS[0].Priority = 0x5;
-        input_report.FINGERS[0].State = 0x0;
-        handleReport(kernel_buffer, kIOHIDReportTypeInput);
+        input_report->FINGERS[0].Priority = 0x5;
+        input_report->FINGERS[0].State = 0x0;
+        handleReport(input_report_buffer,kIOHIDReportTypeInput);
 
         milli_timestamp += 10;
-        input_report.timestamp_buffer[0] = (milli_timestamp << 0x3) | 0x4;
-        input_report.timestamp_buffer[1] = (milli_timestamp >> 0x5) & 0xFF;
-        input_report.timestamp_buffer[2] = (milli_timestamp >> 0xd) & 0xFF;
-        handleReport(kernel_buffer, kIOHIDReportTypeInput);
+        input_report->timestamp_buffer[0] = (milli_timestamp << 0x3) | 0x4;
+        input_report->timestamp_buffer[1] = (milli_timestamp >> 0x5) & 0xFF;
+        input_report->timestamp_buffer[2] = (milli_timestamp >> 0xd) & 0xFF;
+        handleReport(input_report_buffer,kIOHIDReportTypeInput);
     }
-    input_report = {};
+
+    memset(input_report, 0, sizeof(MAGIC_TRACKPAD_INPUT_REPORT));
 }
 
 bool VoodooInputSimulatorDevice::start(IOService* provider) {
     if (!super::start(provider))
         return false;
     ready_for_reports = false;
 
-    kernel_buffer = IOMemoryDescriptor::withAddressRange((mach_vm_address_t)&input_report, sizeof(input_report), kIODirectionNone, kernel_task);
-    if (!kernel_buffer) {
-        IOLog("%s Could not allocate kernel_buffer\n", getName());
+    input_report_buffer = IOBufferMemoryDescriptor::inTaskWithOptions(kernel_task, 0, sizeof(MAGIC_TRACKPAD_INPUT_REPORT));
+    if (!input_report_buffer) {
+        IOLog("%s Could not allocate IOBufferMemoryDescriptor\n", getName());
         return false;
     }
-    kernel_buffer->prepare();
+    input_report = (MAGIC_TRACKPAD_INPUT_REPORT *) input_report_buffer->getBytesNoCopy();
```

Comment:
Why are you doing this? Why not just make it a member of Simulator and use memcpy here?

You don't even release the previous input_buffer. This is a terrible idea as now you're manually managing memory again.

Reply:
> You don't even release the previous input_buffer. This is a terrible idea as now you're manually managing memory again.

Which input buffer are you talking about?  input_report_buffer is allocated once to share with kernel.Then this bufffer is getting reused in handlereport() . This buffer is relased in releaseResource().Single buffer is used throughout this instance.

Why is there a need to memcpy when i can. directly edit memory in buffer? It's the same thing here. **input_report** points to **input_report_buffer**.So we can just change input_report data and submit input_report_buffer to handlereport() since they are pointing to same memory.we don't need extra memcpy here.

Reply:
Ignore my comment, see https://github.com/acidanthera/VoodooInput/pull/14#discussion_r425158558.